### PR TITLE
PLFM-4692 - Refactor team membership invitation and request objects

### PR DESCRIPTION
**Pull request by hhu94**
Changes made:

Renamed `MembershipInvtnSubmission` to `MembershipInvitation`, overwriting the existing `MembershipInvitation`.
Refactored all calls of `MembershipInvitation#getUserId()` to `MembershipInvitation#getInviteeId()`.
Refactored all calls of `MembershipInvitation#setUserId()` to `MembershipInvitation#setInviteeId()`.
Renamed `DBOMembershipInvtnSubmissionDAO` to `MembershipInvitationDAO`.
Renamed `DBOMembershipInvtnSubmissionDAOImpl` to `MembershipInvitationDAOImpl`.
Renamed `DBOMembershipInvtnSubmissionDAOImplTest` to `MembershipInvitationDAOImplTest`.
Renamed `MembershipInvtnSubmissionUtils` to `MembershipInvitationUtils`.
Renamed `DBOMembershipInvtnSubmission` to `DBOMembershipInvitation`.
Renamed `DBOMembershipInvtnSubmissionTest` to `DBOMembershipInvitationTest`.
Renamed `MembershipInvtnSubmission-ddl.sql` to `MembershipInvitation-ddl.sql`.

Renamed `MembershipRqstSubmission` to `MembershipRequest`, overwriting the existing `MembershipRequest`.
Renamed `DBOMembershipRqstSubmissionDAO` to `MembershipRequestDAO`.
Renamed `DBOMembershipRqstSubmissionDAOImpl` to `MembershipRequestDAOImpl`.
Renamed `DBOMembershipRqstSubmissionDAOImplTest` to `MembershipRequestDAOImplTest`.
Renamed `MembershipRqstSubmissionUtils` to `MembershipRequestUtils`.
Renamed `DBOMembershipRqstSubmission` to `DBOMembershipRequest`.
Renamed `DBOMembershipRqstSubmissionTest` to `DBOMembershipRequestTest`.
Renamed `MembershipRqstSubmission-ddl.sql` to `MembershipRequest-ddl.sql`.

Some `AuthorizationManager` and `DAO` methods were renamed in a similar manner (removing the "Submission" part of the name).
Some `DAO` methods were removed because they became redundant as a result of this refactoring.

Note that the database table names (`MEMBERSHIP_INVTN_SUBMISSION` and `MEMBERSHIP_RQST_SUBMISSION`) were left unchanged for fear of breaking migration.

### CODE REVIEW

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMembershipInvitation.java [hunk 1]
```diff
@@ -1,276 +1,300 @@
...
+				if (backup.getEtag() == null) {
+					backup.setEtag(defaultEtag);
+				}
+				try {
+					String backupProperties = new String(unzip(backup.getProperties()));
+					if (backupProperties.contains("MembershipInvtnSubmission")) {
+						// This backup object has a MembershipInvtnSubmission in its properties field
+						// Set its properties to be a MembershipInvitation instead
+						String correctedProperties = backupProperties.replace("MembershipInvtnSubmission", "MembershipInvitation");
+						backup.setProperties(zip(correctedProperties.getBytes()));
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMembershipInvitation.java]: We need to make sure that the result of the transformation is valid.  There are two approaches.  The most conservative approach is:
1) deserialize the properties into a MembershipInvtnSubmission object, the class ONLY being defined in the migrator and used no where else in Synapse.  (Copy the auto-generated POJO here.)
2) Copy fields from the MembershipInvtnSubmission object into the MembershipInvitation object;
3) Serialize the new MembershipInvitation object into the properties field of the DBO.

Another approach is to validate the correctedProperties: call MembershipInvitationUtils.deserialize(correctedProperties.getBytes());

A third approach is to use the name 'aliasing' feature of XStream to directly deserialize the legacy object into a MembershipInvitation object.

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/migration/MigrationType.json [hunk 2]
```diff
@@ -30,8 +30,8 @@
         "DOCKER_REPOSITORY_NAME",
         "DOCKER_COMMIT",
         "TEAM",
-        "MEMBERSHIP_INVITATION_SUBMISSION",
-        "MEMBERSHIP_REQUEST_SUBMISSION",
+        "MEMBERSHIP_INVITATION",
+        "MEMBERSHIP_REQUEST",
```
Comment (brucehoff) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/migration/MigrationType.json]: Check with John and/or Xa:  Are we free to change these names, or will it trip up the migrator.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMembershipRequest.java [hunk 3]
```diff
@@ -1,264 +1,279 @@
...
+		return new MigratableTableTranslation<DBOMembershipRequest, DBOMembershipRequest>(){
+
+			@Override
+			public DBOMembershipRequest createDatabaseObjectFromBackup(DBOMembershipRequest backup) {
+				try {
+					String backupProperties = new String(unzip(backup.getProperties()));
+					if (backupProperties.contains("MembershipRqstSubmission")) {
+						// This backup object has a MembershipRqstSubmission in its properties field
+						// Set its properties to be a MembershipRequest instead
+						String correctedProperties = backupProperties.replace("MembershipRqstSubmission", "MembershipRequest");
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMembershipRequest.java]: See the comments on DBOMembershipInvitation

File: lib/jdomodels/src/main/resources/schema/MembershipInvitation-ddl.sql [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
-  `INVITEE_ID` bigint(20),
-  `INVITEE_EMAIL` varchar(320),
-  `CREATED_ON` bigint(20) NOT NULL,
-  `EXPIRES_ON` bigint(20),
-  `PROPERTIES` mediumblob,
-  PRIMARY KEY (`ID`),
-  CONSTRAINT `MEMBERSHIP_INVITATION_INVITEE_FK` FOREIGN KEY (`INVITEE_ID`) REFERENCES `JDOUSERGROUP` (`ID`) ON DELETE CASCADE,
-  CONSTRAINT `MEMBERSHIP_INVITATION_TEAM_FK` FOREIGN KEY (`TEAM_ID`) REFERENCES `TEAM` (`ID`) ON DELETE CASCADE
-)
+CREATE TABLE IF NOT EXISTS `MEMBERSHIP_INVITATION_SUBMISSION` (
```
Comment (brucehoff) [inline lib/jdomodels/src/main/resources/schema/MembershipInvitation-ddl.sql:1]: Once we've verified that migration works, change the table name to MEMBERSHIP_INVITATION.

File: lib/jdomodels/src/main/resources/schema/MembershipRequest-ddl.sql [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
-  `TEAM_ID` bigint(20) NOT NULL,
-  `CREATED_ON` bigint(20) NOT NULL,
-  `USER_ID` bigint(20) NOT NULL,
-  `EXPIRES_ON` bigint(20),
-  `PROPERTIES` mediumblob,
-  PRIMARY KEY (`ID`),
-  CONSTRAINT `MEMBERSHIP_REQUEST_TEAM_FK` FOREIGN KEY (`TEAM_ID`) REFERENCES `TEAM` (`ID`) ON DELETE CASCADE,
-  CONSTRAINT `MEMBERSHIP_REQUEST_USER_FK` FOREIGN KEY (`USER_ID`) REFERENCES `JDOUSERGROUP` (`ID`)
-  
+CREATE TABLE IF NOT EXISTS `MEMBERSHIP_REQUEST_SUBMISSION` (
```
Comment (brucehoff) [inline lib/jdomodels/src/main/resources/schema/MembershipRequest-ddl.sql:1]: Once we've verified that migration works, change the table name to MEMBERSHIP_REQUEST, please.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMembershipInvitationTest.java [hunk 6]
```diff
@@ -1,137 +1,156 @@
...
+		DBOMembershipInvitation expected = backup;
+		expected.setEtag(DBOMembershipInvitation.defaultEtag);
+		assertEquals(expected, actual);
+		assertEquals(dto, MembershipInvitationUtils.copyDboToDto(dbo));
+	}
+
+	@Test
+	public void testTranslatorRefactor() throws IOException {
+		DBOMembershipInvitation backup = new DBOMembershipInvitation();
+		String oldProperties = "<MembershipInvtnSubmission/>";
```
Comment (brucehoff) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMembershipInvitationTest.java]: Use an actual serialized object from the DB, for example:
```
<MembershipInvtnSubmission>
  <createdOn>2017-10-30 15:54:48.793 UTC</createdOn>
  <message></message>
  <id>9603838</id>
  <createdBy>273995</createdBy>
  <inviteeId>3341743</inviteeId>
  <teamId>3320424</teamId>
</MembershipInvtnSubmission>
```

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMembershipRequestTest.java [hunk 7]
```diff
@@ -1,107 +1,125 @@
...
+		params = new MapSqlParameterSource();
+		params.addValue("id", clone.getId());
+		DBOMembershipRequest clone2 = dboBasicDao.getObjectByPrimaryKey(DBOMembershipRequest.class, params);
+		assertEquals(clone, clone2);
+	}
+
+	@Test
+	public void testTranslatorRefactor() throws IOException {
+		DBOMembershipRequest backup = new DBOMembershipRequest();
+		String oldProperties = "<MembershipRqstSubmission/>";
```
Comment (brucehoff) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMembershipRequestTest.java]:
```
<MembershipRqstSubmission>
  <createdOn>2014-02-05 22:21:37.765 UTC</createdOn>
  <createdBy>1976831</createdBy>
  <userId>2223382</userId>
  <teamId>2223746</teamId>
</MembershipRqstSubmission>
```

File: lib/jdomodels/src/main/resources/schema/MembershipInvitation-ddl.sql (hunk 4 above)
Comment (hhu94) [reply lib/jdomodels/src/main/resources/schema/MembershipInvitation-ddl.sql:1]: Tracking this task in https://sagebionetworks.jira.com/browse/PLFM-4716

File: lib/jdomodels/src/main/resources/schema/MembershipRequest-ddl.sql (hunk 5 above)
Comment (hhu94) [reply lib/jdomodels/src/main/resources/schema/MembershipRequest-ddl.sql:1]: Tracking this task in https://sagebionetworks.jira.com/browse/PLFM-4716

**Pull request merged**